README.md

```diff
@@ -115,6 +117,50 @@ The promise **must** return a response object anything else will lead to a 'HTTP
 
 Other types aren't allowed and will lead to a 'HTTP 500 Internal Server Error' response for the client.
 
+### Middleware
+
+#### Creating your own middleware
+
+You can create your own middleware, which lies between the server and the callback function. Use this middleware to manipulate the requests or
+responses of the server. You can add as many middleware as you want you just need to follow the following design
+
+```php
+$callback = function (RequestInterface $request) {
+    return new Response();
+}
+
+$middleware = function (RequestInterface $request, callable $next) {
+    // check or maninpulate the request object
+    ...
+    // call of next middleware chain link
+    return $next($request);
+}
+
+$server = new HttpServer($socket, $callback);
+$server->addMiddleware($middleware);
+```
+
+Make sure you add the `return $next($request)` in your middleware code. Otherwise the client will receive a `500 Internal Server Error` message.
```

Comment:
I suppose this is a bit unclear. There are valid cases where `$next` will *not* be called, right?